components/places/src/storage/bookmarks.rs

```diff
@@ -376,6 +395,259 @@ fn delete_bookmark_in_tx(db: &impl ConnExt, guid: &SyncGuid) -> Result<bool> {
     Ok(true)
 }
 
+/// Support for modifying bookmarks, including changing the location in
+/// the tree.
+
+/// Used instead of Option<String> for updating the title, so we can
+/// differentiate between "no change", "set to null" and "set to a value"
+/// Could trivially use <T>, but title is the only use-case for now, so it's
+/// a little clearer to leave it specific.
+#[derive(Debug, Clone)]
+pub enum UpdateTitle {
```

Comment:
Hmm, in [bug 1360872](https://bugzilla.mozilla.org/show_bug.cgi?id=1360872), we removed the distinction between `""` and `null`, as it was confusing and not really useful. So empty strings get stored as SQL `NULL`s in the database, and get returned as `null`, regardless of whether you passed in `""` or `null` originally.

Could we use a plain `Option` here, with `None` meaning "don't update", `Some("new title")` to mean "change", and `Some("")` or `Some(String::new())` to mean "clear"? Neither `""` nor `String::new()` should allocate.